runtime/fs_node.js

```diff
@@ -104,16 +121,161 @@ MlNodeDevice.prototype.open = function(name, f) {
   } catch (err) {
-    caml_raise_sys_error(err.toString());
+    if (raise_unix) {
+      this.raise_unix_exn_of_nodejs_error(err);
+    } else {
+      caml_raise_sys_error(err.toString());
+    }
+  }
+}
+MlNodeDevice.prototype.stat = function(name, raise_unix) {
+  try {
+    var js_stats = this.fs.statSync(this.nm(name));
+    return this.stats_from_js(js_stats);
+  } catch (err) {
+    if (raise_unix) {
+      this.raise_unix_exn_of_nodejs_error(err);
+    } else {
+      caml_raise_sys_error(err.toString());
+    }
+  }
+}
+MlNodeDevice.prototype.lstat = function(name, raise_unix) {
+  try {
+    var js_stats = this.fs.lstatSync(this.nm(name));
+    return this.stats_from_js(js_stats);
+  } catch (err) {
+    if (raise_unix) {
+      this.raise_unix_exn_of_nodejs_error(err);
+    } else {
+      caml_raise_sys_error(err.toString());
+    }
+  }
+}
+MlNodeDevice.prototype.symlink = function(to_dir, src, dst, raise_unix) {
+  try {
+    this.fs.symlinkSync(this.nm(src), this.nm(dst), to_dir ? 'dir' : 'file');
+    return 0;
+  } catch (err) {
+    if (raise_unix) {
+      this.raise_unix_exn_of_nodejs_error(err);
+    } else {
+      caml_raise_sys_error(err.toString());
+    }
+  }
+}
+MlNodeDevice.prototype.readlink = function(name, raise_unix) {
+  try {
+    var js_stats = this.fs.readlinkSync(this.nm(name), 'utf8');
+    return caml_string_of_jsbytes(js_stats);
+  } catch (err) {
+    if (raise_unix) {
+      this.raise_unix_exn_of_nodejs_error(err);
+    } else {
+      caml_raise_sys_error(err.toString());
+    }
+  }
+}
+MlNodeDevice.prototype.raise_unix_exn_of_nodejs_error = function(err) {
+  /* ===Unix.error===
+   *
+   * This array is in order of the variant in OCaml
+   */
+  var errors = [
+    "E2BIG", "EACCES", "EAGAIN", "EBADF", "EBUSY", "ECHILD", "EDEADLK", "EDOM",
+    "EEXIST", "EFAULT", "EFBIG", "EINTR", "EINVAL", "EIO", "EISDIR", "EMFILE",
+    "EMLINK", "ENAMETOOLONG", "ENFILE", "ENODEV", "ENOENT", "ENOEXEC", "ENOLCK",
+    "ENOMEM", "ENOSPC", "ENOSYS", "ENOTDIR", "ENOTEMPTY", "ENOTTY", "ENXIO",
+    "EPERM", "EPIPE", "ERANGE", "EROFS", "ESPIPE", "ESRCH", "EXDEV", "EWOULDBLOCK",
+    "EINPROGRESS", "EALREADY", "ENOTSOCK", "EDESTADDRREQ", "EMSGSIZE",
+    "EPROTOTYPE", "ENOPROTOOPT", "EPROTONOSUPPORT", "ESOCKTNOSUPPORT",
+    "EOPNOTSUPP", "EPFNOSUPPORT", "EAFNOSUPPORT", "EADDRINUSE", "EADDRNOTAVAIL",
+    "ENETDOWN", "ENETUNREACH", "ENETRESET", "ECONNABORTED", "ECONNRESET", "ENOBUFS",
+    "EISCONN", "ENOTCONN", "ESHUTDOWN", "ETOOMANYREFS", "ETIMEDOUT", "ECONNREFUSED",
+    "EHOSTDOWN", "EHOSTUNREACH", "ELOOP", "EOVERFLOW"
+  ];
+  var variant = errors.indexOf(err.code);
+  if (variant < 0) {
+    // If none of the above variants, fallback to EUNKNOWNERR(int)
+    variant = BLOCK(0, err.errno);
+  }
+  var args = [
+    variant,
+    caml_string_of_jsbytes(err.syscall || ""),
+    caml_string_of_jsbytes(err.path || "")
+  ];
+  caml_raise_with_args(caml_named_value("Unix.Unix_error"), args.length, args);
```

Comment:
We might want to have a fallback in case `Unix.Unix_error` was not registered.

Reply:
Is there a way to determine if `Unix.Unix_error` was registered inside the JS?

Reply:
`if(caml_named_value("Unix.Unix_error"))`

Reply:
Is it fine to fallback to the Sys_error if `Unix.Unix_error` is not registered? I changed the `if` statement to `if (raise_unix && caml_named_value("Unix.Unix_error")) {`

Reply:
looks good. Maybe make a single call to caml_named_value.